sklearn/preprocessing/imputation.py

```diff
@@ -413,6 +431,21 @@ def transform(self, X):
         valid_idx = np.where(valid_mask)[0]
         missing = np.arange(X.shape[not self.axis])[invalid_mask]
 
+        if sparse.issparse(X) and self.missing_values != 0:
+            # sparse matrix and missing values is not zero
+            mask = _get_mask(X.data, self.missing_values)
+            indexes = np.repeat(np.arange(len(X.indptr) - 1, dtype=np.int),
+                                np.diff(X.indptr))[mask]
+            self.mask_matrix_ = X.__class__((mask, X.indices.copy(),
+                                             X.indptr.copy()), shape=X.shape,
+                                            dtype=X.dtype)
+            self.mask_matrix_.eliminate_zeros()  # remove explicit 0 entries
+        else:
+            # sparse with zero as missing value and dense matrix
+            if sparse.issparse(X):
+                X = X.toarray()
+            self.mask_ = _get_mask(X, self.missing_values)
+
         if self.axis == 0 and invalid_mask.any():
```

Comment:
A simple way to fix the code structure is to put this in a separate helper and call it from each of `_{sparse,dense}_transform`